core/client/fs/src/main/java/alluxio/client/block/stream/GrpcBlockingStream.java

```diff
@@ -145,14 +148,56 @@ public void send(ReqT request) throws IOException {
    * @throws IOException if any error occurs
    */
   public ResT receive(long timeoutMs) throws IOException {
+    return retryReceive(timeoutMs, null);
+  }
+
+  /**
+   * Receives a response from the server. Will wait until a response is received, or
+   * throw an exception if times out.
+   *
+   * @param timeoutMs maximum time to wait before giving up and throwing
+   *                  a {@link DeadlineExceededException}
+   * @param request   the request
+   * @return the response message, or null if the inbound stream is completed
+   * @throws IOException if any error occurs
+   */
+  public ResT retryReceive(long timeoutMs, ReqT request) throws IOException {
     if (mCompleted) {
       return null;
     }
     if (mCanceled) {
       throw new CancelledException(formatErrorMessage("Stream is already canceled."));
     }
     try {
-      Object response = mResponses.poll(timeoutMs, TimeUnit.MILLISECONDS);
+      Object response;
+      if (request == null) {
+        response = mResponses.poll(timeoutMs, TimeUnit.MILLISECONDS);
+      } else {
+        Clock clock = Clock.systemUTC();
+        Duration maxWaitTime = Duration.ofSeconds(3);
+        Duration nextWaitTime = Duration.ofMillis(32);
```

Comment:
Do you think the `maxWaitTime` or `nextWaitTime` should be configurable?

Also, have you considered using something from the `RetryUtils` class? It should make the code easier to read while keeping similar behavior.

Reply:
Also, the default value passed for the `timeoutMs` is equal to the value `alluxio.user.streaming.data.timeout` which defaults to 30s. Since it is possible to configure the value higher or lower, I think rather than make it user-configurable I would rather just dynamically adjust the `maxWaitTime` based on the `timeoutMs` parameter.

What do you think?